### PR TITLE
[2.0.x] Add support for LPC1769 at 120 MHz

### PR DESCRIPTION
**Pull request by tcm0116**
I've been working on getting LPC1769 running at 120 MHz, so I decided to go ahead and share where I'm currently at. I'll continue to update this PR as I work toward getting the Cohesion3D ReMix board I've got incorporated into one of my printers.

### CODE REVIEW

**Comment by cjsoong** [comment]
Can MKS Sbase read the onboard SD card?
How to set up using the onboard SD card?

**Comment by tcm0116** [comment]
@cjsoong Marlin currently doesn't support reading G-code files from the onboard SD card.

**Comment by Bob-the-Kuhn** [comment]
The ST7920 is finicky about it's timing.  With the hardware SPI it'll run at 2MHz.  Some software SPIs will only get 500KHz out of it.

In **u8g_com_HAL_LPC1768_st7920_sw_spi.cpp** try changing the **#define SPI_SPEED 4** to 5 and see if that fixes the RRD Full Graphic LCD issue.

**Comment by tcm0116** [comment]
@Bob-the-Kuhn It's all good now. I re-worked the Software SPI in the LPC1768 HAL to provide a single implementation that is shared amongst the SD card and the two U8G software SPI bindings. I used a scope to characterize the performance of my implementation at both 100 MHz and 120 MHz, and can get a solid 1 MHz. Please have a look and let me know if you have any concerns.

**Comment by tcm0116** [comment]
@Bob-the-Kuhn do you happen to have a ST7565 LCD that you could test my changes with?

**Comment by Bob-the-Kuhn** [comment]
I have a VIKI2 which is ST 7565 based.  I should be able to test it today.

**Comment by Bob-the-Kuhn** [comment]
My Vicki2 works with your software.  

The only change I had to make was in the **pins_RAMPS_RE_ARM.h** file.  Someone added MK2 Multiplexer Support but never tested it with the default configuration.   I have a PR under review that'll fix that.

**Comment by tcm0116** [comment]
Thanks for testing that. How did you have the VIKI2 connected? If it's connected to the hardware SPI pins and SDSUPPORT is disabled, it should use the hardware SPI. If SDSUPPORT is enabled or if the LCD is not connected to the hardware SPI pins, it should use the software SPI. 

If we added support for `spiBeginTransaction` to the LPC1768 HAL, then we should be able to call that from the SD card code and the U8G HAL bindings, and be able to use hardware SPI if the two devices share a hardware SPI channel. All of the other Arduino-based HALs should already have `spiBeginTransaction`, so it shouldn't break anything.

**Comment by tcm0116** [comment]
I've removed **[Experimental]** from the title as I'm now satisfied that these changes are ready to merge. I've successfully completed a print on one of my machines running a Cohesion3D ReMix board (LPC1769) equipped with TMC2130s (SPI interface), a BLTouch, and a RepRap Discount Full Graphic LCD.

The ReMix board brings out the SPI interface for the onboard SD card to the LCD controller, so I wasn't able to verify concurrent use of the SPI bus to access the LCD and remote SD card. However, I ran these changes on my Re-ARM with the RRD Full Graphic Controller, and did verify that the concurrent SPI access works on the LPC1768 at 100 MHz.

I verified the new software SPI implementation using an oscilloscope and was able to tune the speed of the implementation to closely match the SPI speeds used by Marlin for the hardware SPI up to the limits of the software SPI implementation (5 MHZ for LPC1768 and 6 MHz for LPC1769). At and below 2 MHz, the speed of the software SPI is very close to the desired speed on both MCUs. The became evident in my testing with the Full Graphic LCD as well as @Bob-the-Kuhn's testing using a VIKI2 LCD.

With these changes, we can also support Smoothieboard v1 by with the addition of the necessary pins file. The Cohesion3D ReMix is basically a Smoothieboard v1 clone with a few additions, so the pins file I'm working on for that can easily be used as a basis for the Smootheboard pins file.

File: Marlin/src/HAL/HAL_LPC1768/HAL_spi.cpp [hunk 1]
```diff
@@ -47,113 +47,41 @@
  * https://github.com/MarlinFirmware/Marlin/tree/071c7a78f27078fd4aee9a3ef365fcf5e143531e
  */
 
+#include "src/inc/MarlinConfig.h"
```
Comment (thinkyhead) [inline Marlin/src/HAL/HAL_LPC1768/HAL_spi.cpp]: This works?

But it should still be a relative path.

File: Marlin/src/HAL/HAL_LPC1768/SoftwareSPI.h [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+   *     6 - about 125 kHz peak
+   */
+
+  void swSpiBegin(pin_t sck_pin, pin_t miso_pin, pin_t mosi_pin);
+
+  // Returns the spi_speed value to be passed to swSpiTransfer
+  uint8_t swSpiInit(uint8_t spiRate, pin_t sck_pin, pin_t mosi_pin);
+
+  uint8_t swSpiTransfer(uint8_t b, uint8_t spi_speed, pin_t sck_pin, pin_t miso_pin, pin_t mosi_pin);
+
```
Comment (thinkyhead) [inline Marlin/src/HAL/HAL_LPC1768/SoftwareSPI.h]: For 2.0 we're not indenting the content between the outer-most conditions in headers and other files that contain a single encompassing condition.

**Comment by Bob-the-Kuhn** [comment]
#include "src/inc/MarlinConfig.h"

Works in PlatformIO but not Arduino.

I'll do a PR to fix both places.

**Comment by tcm0116** [comment]
I wasn't aware that it doesn't work in Arduino. Thanks!

**Comment by Bob-the-Kuhn** [comment]
Scott beat us to it.

**Comment by thinkyhead** [comment]
And Bob got the rest!

**Comment by Bob-the-Kuhn** [comment]
@tcm0116 - why the `DISABLED(SDSUPPORT) ` in ultralcd_impl_DOGM.h?

Some systems use the hardware SPI for both the SD card and the LCD.

I'd like to remove that.

I think the` (DOGLCD_SCK == SCK_PIN) && (DOGLCD_MOSI == MOSI_PIN)` parts are a good idea.

**Comment by tcm0116** [comment]
>why the DISABLED(SDSUPPORT) in ultralcd_impl_DOGM.h?

Because the LCD and SD card operate at different speeds and modes, and the U8G library doesn't currently change the speed or mode before starting a SPI transaction. As such, the first SD card operation will increase the speed to 8 MHz, and it will stay there with will cause the LCD to be garbled.

We can remove that restriction once:

- U8G sets the mode and speed when starting a SPI transaction
- The `Sd2Card` class sets the mode at the start of a transaction
- The LPC hardware SPI implementation supports changing the mode

**Comment by Bob-the-Kuhn** [comment]
Looks like we have some more work to do.

My memory is a bit rusty on the details.


> We can remove that restriction once:
> - U8G sets the mode and speed when starting a SPI transaction
> - The Sd2Card class sets the mode at the start of a transaction
> - The LPC hardware SPI implementation supports changing the mode

I'm not aware of any of the spiInit() routines that support mode changes.

For sure the SD Card code sets the SPI rate at the beginning of a new transaction.

At least one of the U8G routines does NOT set the SPI speed at the beginning of a new transaction.

What will stop working if DISABLED(SDSUPPORT) is removed?

----

I've verified this afternoon that the AVR boards work just fine when using the hardware SPI for both the SD card and the LCD when using VIKI2.

What isn't working is AVR with software SPI for the LCD and hardware SPI for the SD card when both use the same pins.  I need to get that working again for the VIKI2.  So far no luck in even proving which LCD COM routine is being used.

I'm 99.9% sure that the AVR LCD software SPI was broken before this PR.  

----

What we really need is to implement a SPI manager and bring all the various SPI users into compliance.

I had one working for the LPC1768.  The next step was to see if it could be extended to the AVR boards.  If I continue to feel better I'll re-start that in a few days using an older copy of 2.0.x

**Comment by tcm0116** [comment]
>I'm not aware of any of the spiInit() routines that support mode changes.

That's where `beginTransaction` comes into play using the `SPIClass`.

>What will stop working if DISABLED(SDSUPPORT) is removed?

The RRD Full Graphic LCD Controller (which is the U8G binding in question) because the SD card will set the speed to 8 MHz, which is too fast for the 12864 controller.

>What isn't working is AVR with software SPI for the LCD and hardware SPI for the SD card when both use the same pins.

I'm 99.9% sure that would never work because I doubt you can toggle a pin manually that's been configured as a hardware SPI pin.

>What we really need is to implement a SPI manager and bring all the various SPI users into compliance.

Probably, but what I've done for LPC1768 seems to work pretty well with the software SPI backend that allows pins to be specified and the standard hear ware SPI interface that uses the single SPI channel available (the other is used by the onboard SD card).

For now, I was just trying to take a step in the right direction. As it was before, the LCD on LPC1768 was always relegated to using software SPI, which is really inefficient. For now, the hardware SPI will be used if it's not shared, but that can be easily fixed by making the `Sd2Card` class and the U8G SPI bindings play nicely together.

**Comment by Bob-the-Kuhn** [comment]
> I'm 99.9% sure that would never work because I doubt you can toggle a pin manually that's been configured as a hardware SPI pin.

Used to work for AVR.

Definitely doesn't work for LPC176x.

----

> The RRD Full Graphic LCD Controller

It's been a **REAL PAIN**.  That's the whole reason for keeping the LPC176x as a software SPI.

----

We need to come up with something that works for both the LPC176xx boards and the AVR boards.

We need a quick (1 or 2 day) solution.  

My first thought is to put the LPC176x back into shared SPI mode all the time and remove the DISABLED(SDSUPPORT) until we have a better solution.

Time to wind down for the night.   Let me know your ideas for a short term solution.

**Comment by tcm0116** [comment]
>Used to work for AVR.

I was of the impression that on AVR, the SD card used a hardware SPI channel and the LCD used a software SPI channel using different pins.

>My first thought is to put the LPC176x back into shared SPI mode all the time and remove the DISABLED(SDSUPPORT) until we have a better solution.

It already reverts to shared software SPI mode if both the 12864 LCD and the SD Card is enabled and using the same pins. The only change from how it was before this PR is that if you only have one of those devices on the hardware SPI bus, then it will go ahead and use the hardware SPI. I'm not aware of anything that has been broken that would require an immediate solution. Am I missing something?

**Comment by Bob-the-Kuhn** [comment]
It's a challenge to keep all the various bits & pieces playing nicely.

The only things I know for sure are:
-   AVR VIKI2 is broken on 1.1.x
-   AVR VIKI2 is broken on 2.0.x but will work if the DISABLED(SDSUPPORT) is removed 

The VIKI2 shares the same SPI for the LCD and the SD card.  Most others have separate SPIs for the LCD and the SD card.  

Hopefully I'll have some time tomorrow to work on this.

Good job on this PR.   It'll be a very pleasant surprise if this is the only unintended item that comes up.

**Comment by tcm0116** [comment]
Removing `DISABLED(SDSUPPORT)` will cause Marlin to use hardware SPI for the LCD. Before this PR was merged, Marlin would use software SPI for the LCD and probably hardware SPI for the SD card. I really don't think you can use software and hardware SPI on the same pins.

If I had to guess, this commit is where the VIKI2 was broken: https://github.com/MarlinFirmware/Marlin/commit/381ebc043fe0cba6f915521114afefe633e7b384#diff-695a0b5d0447a1f65455a945b7de113d

My guess is that the ST7565 controller can handle the 8 MHz SPI speed that the `Sd2Card` class uses for the hardware SPI, which is why it's working on 2.0 with `DISABLED(SDSUPPORT)` removed. If that's the case, then removing `DISABLED(SDSUPPORT)` is definitely a viable option for the ST7565, and then you'll need to figure out how to get it to use hardware SPI on 1.1.x.

>Good job on this PR. It'll be a very pleasant surprise if this is the only unintended item that comes up.

Thanks. I don't think this issue was caused by my PR. However, it does look like it may have helped to find a solution.

**Comment by thinkyhead** [comment]
It's been ages since I was using a VIKI2 (with an MKS_13 board). It was working well at that time, so should definitely be fixable.

**Comment by Bob-the-Kuhn** [comment]
I was wrong about the soft and hard SPIs peacefully coexisting on the AVR CPUs.

I've finally gotten VIKI2 to work on bugfix-1.1.x.  

The only places (that I know of) where the LCD and the SD card share pins are the VIKI type displays and the RE_ARM board.

Removing the DISABLED(SDSUPPORT) from just the ST7565 is reasonable.  The ST7565 can run at the SD card rates with no problems.

I also found that the **pins_RAMPS_RE_ARM.h** needed `#define SDSS  P1_23` added to it.  Apparently the SDSS definition in **pins.h** is being processed before the one in **spi_pins.h** which resulted in not being able to read the SD card.

I need to check Due to see if it has the same SDSS issue.

**Comment by Bob-the-Kuhn** [comment]
I've just been reminded that the MKS mini12864 also has a shared SPI.

File: frameworks/CMSIS/LPC1768/Re-ARM/system_LPC17xx.c [hunk 3]
```diff
@@ -546,9 +558,15 @@ void SystemInit (void)
 
   LPC_SC->CCLKCFG   = 0x00000002;       /* Setup CPU Clock Divider            */
 
-  LPC_SC->PLL0CFG   = 0x00010018;     // 100MHz
-  LPC_SC->PLL0FEED  = 0xAA;
-  LPC_SC->PLL0FEED  = 0x55;
+  if(can_120MHz()) {
+    LPC_SC->PLL0CFG   = 0x0000000E;     /* configure PLL0                     */
+    LPC_SC->PLL0FEED  = 0xAA;
+    LPC_SC->PLL0FEED  = 0x55;
```
Comment (Artem-B) [inline frameworks/CMSIS/LPC1768/Re-ARM/system_LPC17xx.c:564]: We also need to adjust flash access time in FLASHCFG.FLASHTIM as the default is too aggressive for 120MHz. It's currently set to 4 cycles and it should be 5 for 120MHz 1769. [Ref: section 5.4 in the manual](https://www.nxp.com/docs/en/user-guide/UM10360.pdf)

**Pull request merged**